elastic/kibana
# [Runtime field editor] Preview field against cluster data

This PR adds the previewing of the runtime field in the context of the latest documents from the index pattern.

When we open the index pattern field editor we load the latest 50 documents from the cluster and display in the Preview panel their fields (initially we only display the first 7 fields).

The error handling currently in place is basic and will be improved on as a separate chunk of work.

## Notes on implementation

I have used a virtual list to render the list of fields. This allows the UI to support huge list of fields (5000+) that can occur in certain scenarios. This implies that the list items must have a fix height and that we give an height to the scrollable area.
The height is calculated dynamically with a `<EuiResizeObserver />` and passed down to the `<VirtualList />` component so we can make use of the full height of the flyout panel to render the fields on any screen size.

## How to test

* Make sure to load the Kibana sample data logs
* Navigate to the index pattern management app and select the "kibana_sample_data_logs" index pattern
* Click "Add field"
* Give a name to the field and toggle the "Set value"
* The Preview pannel should open with the empty prompt
* Add a script (e.g. `emit("hello")`)
* The Preview panel should update and display the runtime field (name and value) along with the other documents fields
* Click the "Show more" button to display all the fields
* The fields should expand and you should be able to scroll all the fields.
* Now let's do some calculation with our script: change the field type to "long" and the script to `emit(doc['bytes'].value * 2)`
* The preview should correctly show the double that what the "bytes" field indicates.

* Change the document ID and provide a wrong ID
* There should be an error
* Now enter a valid ID (e.g. "pyDLo3kBEdBikFTNyctP")
* You should see that document in the preview and the "Previous" / "Next" button should not be visible anymore
* Click on the "Load documents from cluster" button
* You should be back navigating the list of documents loaded from the cluster

## Test with 5000 fields

As we are using a virtual list, the number of fields do not have a performance impact in the rendering.  
Open the `field_list.tsx` file and replace the `fieldsValues` declaration block on L49 with

```js
const fieldsValues = useMemo(
    () =>
      new Array(5000).fill('').map((_, i) => ({
        key: `key__${i}`,
        value: `value__${i}`,
      })),
    []
  );
```

## Changes to the Core public API

This is a **minor** change.

I have added the `hideCloseButton` EuiFlyoutProp to be able to hide the close button in the flyout when using the `overlays.openFlyout()` core handler.

## Upcoming work to be done

- [ ] Search fields & pin fields at the top of the list
- [ ] Preview the "Format" that user might define in the editor
- [ ] Component integration tests

## Screenshots

<img width="1670" alt="Screenshot 2021-06-04 at 17 11 52" src="https://user-images.githubusercontent.com/2854616/121027384-d4918980-c79e-11eb-9bd8-e98470c1ad34.png">
<img width="1179" alt="Screenshot 2021-06-04 at 17 12 56" src="https://user-images.githubusercontent.com/2854616/121027401-d8251080-c79e-11eb-8539-ff57cddec010.png">
